modules/lang-painless/src/main/java/org/elasticsearch/painless/PainlessScriptEngine.java

```diff
@@ -130,10 +139,78 @@ public boolean needsScores() {
             };
             return context.factoryClazz.cast(factory);
         } else if (context.instanceClazz.equals(ExecutableScript.class)) {
+            GenericElasticsearchScript painlessScript =
+                (GenericElasticsearchScript)compile(contextsToCompilers.get(context), scriptName, scriptSource, params);
+
             ExecutableScript.Factory factory = (p) -> new ScriptImpl(painlessScript, p, null, null);
             return context.factoryClazz.cast(factory);
+        } else {
+            // Check we ourselves are not being called by unprivileged code.
+            SpecialPermission.check();
+
+            // Create our loader (which loads compiled code with no permissions).
+            final Loader loader = AccessController.doPrivileged(new PrivilegedAction<Loader>() {
+                @Override
+                public Loader run() {
+                    return new Loader(getClass().getClassLoader());
+                }
+            });
+
+            compile(contextsToCompilers.get(context), loader, scriptName, scriptSource, params);
+
+            return generateFactory(loader, context);
+        }
+    }
+
+    private <T> T generateFactory(Loader loader, ScriptContext<T> context) {
```

Comment:
Can you please add a javadoc here giving a high level overview of what is generated (ie ctor with which params, members, etc)?